core/flow/tc_adaptive.go

```diff
@@ -0,0 +1,45 @@
+package flow
+
+import "github.com/alibaba/sentinel-golang/core/system_metric"
+
+// MemoryAdaptiveTrafficShapingCalculator is a memory adaptive traffic shaping calculator
+//
+// adaptive flow control algorithm
+// If the watermark is less than Rule.LowWaterMark, the threshold is Rule.SafeThreshold.
+// If the watermark is greater than Rule.HighWaterMark, the threshold is Rule.RiskThreshold.
+// Otherwise, the threshold is ((watermark - LowWaterMark)/(HighWaterMark - LowWaterMark)) *
+//	(RiskThreshold - SafeThreshold) + SafeThreshold.
+type MemoryAdaptiveTrafficShapingCalculator struct {
+	owner         *TrafficShapingController
+	safeThreshold int64
+	riskThreshold int64
+	lowWaterMark  int64
+	highWaterMark int64
+}
+
+func NewMemoryAdaptiveTrafficShapingCalculator(owner *TrafficShapingController, r *Rule) *MemoryAdaptiveTrafficShapingCalculator {
+	return &MemoryAdaptiveTrafficShapingCalculator{
+		owner:         owner,
+		safeThreshold: r.SafeThreshold,
+		riskThreshold: r.RiskThreshold,
+		lowWaterMark:  r.LowWaterMark,
+		highWaterMark: r.HighWaterMark,
+	}
+}
+
+func (m *MemoryAdaptiveTrafficShapingCalculator) BoundOwner() *TrafficShapingController {
+	return m.owner
+}
+
+func (m *MemoryAdaptiveTrafficShapingCalculator) CalculateAllowedTokens(_ uint32, _ int32) float64 {
+	var threshold float64
+	mem := system_metric.CurrentMemoryUsage()
```

Comment:
We should check if we get CurrentMemoryUsage.

Reply:
in system_metric module, we need another PR to enhance.
Could you please file a Issue?

Reply:
ok